lib/strzero.c

```diff
@@ -0,0 +1,95 @@
+/***************************************************************************
+ *                                  _   _ ____  _
+ *  Project                     ___| | | |  _ \| |
+ *                             / __| | | | |_) | |
+ *                            | (__| |_| |  _ <| |___
+ *                             \___|\___/|_| \_\_____|
+ *
+ * Copyright (C) Daniel Stenberg, <daniel@haxx.se>, et al.
+ *
+ * This software is licensed as described in the file COPYING, which
+ * you should have received as part of this distribution. The terms
+ * are also available at https://curl.se/docs/copyright.html.
+ *
+ * You may opt to use, copy, modify, merge, publish, distribute and/or sell
+ * copies of the Software, and permit persons to whom the Software is
+ * furnished to do so, under the terms of the COPYING file.
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
+ * KIND, either express or implied.
+ *
+ * SPDX-License-Identifier: curl
+ *
+ ***************************************************************************/
+
+#include "curl_setup.h"
+
+#include "strzero.h"
+
+/*
+ * Curl_explicit_bzero zeroes out a memory buffer in such a way to ensure that
+ * it won't be optimized away by the compiler due to being a dead-store. If no
+ * API for this exists it will fall back to plain memset using a volatile ptr.
+ */
+
+#if defined(_WIN32)
+#include <windows.h>
+void
+Curl_explicit_bzero(void *buffer, size_t length)
+{
+  if(!buffer)
+    return;
+  /*
+   * TODO: Windows 11 ships a new function SecureZeroMemory2 which is a safer
+   * method, but is not as widely available as SecureZeroMemory which exist
+   * all the way back to Windows XP.
+   */
+  (void)SecureZeroMemory(buffer, length);
```

Comment:
According to this copy of the SDK:
https://github.com/hughbe/windows-sdk-headers/blob/89f151d343587fcd5b854a4851fcb8f7187f3ac4/Microsoft.Windows.SDK.CPP/latest/Include/um/WinBase.h#L123
a simple `#ifdef` might be enough to detect it, at least with MSVC. We'd need to test it before adding it.

Reply:
Hmm, I agree that it seems like an ifdef is all we need. I've added it to the code in this rebase for a test to see if any blows up.